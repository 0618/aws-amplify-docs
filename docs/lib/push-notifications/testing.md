---
title: Testing
description: description
---

<inline-fragment platform="js" src="~/lib/push-notifications/fragments/js/testing.md"></inline-fragment>
