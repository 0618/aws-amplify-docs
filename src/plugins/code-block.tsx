// eslint-disable-next-line @typescript-eslint/no-var-requires
const prism = require('prismjs');
// eslint-disable-next-line @typescript-eslint/no-var-requires
const loadLanguages = require('prismjs/components/');
// eslint-disable-next-line @typescript-eslint/no-var-requires
const Html5Entities = require('html-entities');
// eslint-disable-next-line @typescript-eslint/no-var-requires
const unified = require('unified');
// eslint-disable-next-line @typescript-eslint/no-var-requires
const rehypeParse = require('rehype-parse');
// eslint-disable-next-line @typescript-eslint/no-var-requires
const versions = require('../constants/versions.ts');

const entities = new Html5Entities.Html5Entities();

const supportedLanguages = [
  'markup',
  'objectivec',
  'html',
  'xml',
  'css',
  'docker',
  'go',
  'ini',
  'js',
  'ts',
  'bash',
  'swift',
  'kotlin',
  'python',
  'java',
  'yaml',
  'ruby',
  'wasm',
  'rust',
  'json',
  'typescript',
  'javascript',
  'graphql',
  'diff',
  'jsx',
  'sql',
  'groovy',
  'dart'
];

loadLanguages(supportedLanguages);

let lineCountOffSet = 0;

const highlight = (code, language, regions, idx) => {
  language = language.replace('language-', '');
  let highlighted = '';
  const languageIsSet = !!(language && language.trim().length > 0);
  code = code.trim();

  if (languageIsSet && prism.languages[language]) {
    if (!supportedLanguages.includes(language)) {
      throw new Error(
        `No support for ${language} syntax highlighting. Contact Amplify JS team to request support.`
      );
    }

    highlighted = prism.highlight(code, prism.languages[language], language);
  } else {
    highlighted = entities.encode(code);
  }

  const html = `<div slot="content" class="highlight highlight-source${
    languageIsSet ? `-${language}` : ''
  }">${highlighted}</div>`;

  let lineCount = html.split(/\r\n|\r|\n/).length;

  if (regions > 1 && idx > 0) {
    lineCountOffSet = lineCountOffSet + lineCount;
  } else {
    lineCountOffSet = 0;
  }

  // const noCopy = code.indexOf('###BEGIN_COPY###') > -1;

  return [
    {
      type: 'element',
      tagName: 'p',
      properties: {
        class: 'searchable-code'
      },
      children: [
        {
          type: 'text',
          value: `${entities.encode(code)}`
        }
      ]
    },
    {
      type: 'jsx',
      value: `<CodeBlock language="${language}" lineCount="${lineCount}" lineCountOffset="${lineCountOffSet}" noCopy="${false}" >`
    },
    ...unified()
      .use(rehypeParse, { fragment: true })
      .parse(html).children,
    {
      type: 'jsx',
      value: `</CodeBlock>`
    }
  ];
};

const addVersions = (code) => {
  code = code.replace(/ANDROID_VERSION/g, versions.ANDROID_VERSION);
  code = code.replace(
    /ANDROID_KOTLIN_VERSION/g,
    versions.ANDROID_KOTLIN_VERSION
  );
  code = code.replace(/ANDROID_GEO_VERSION/g, versions.ANDROID_GEO_VERSION);

  return code;
};

const removeCopy = (code) => {
  code = code.split('###BEGIN_COPY###');
  return code;
};

const reducer = (previousValue, currentValue) =>
  previousValue.concat(currentValue);

const codeBlockPlugin = () => (tree) => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const visit = require('unist-util-visit');

  visit(tree, (node) => {
    if (node.tagName === 'code') {
      let code = addVersions(node.children[0].value);
      code = removeCopy(code);

      const language =
        'className' in node.properties
          ? node.properties.className[0]
          : 'markup';
      node.children = code
        .map((section, idx) => highlight(section, language, code.length, idx))
        .reduce(reducer);

      console.log(node.children);
    }
  });
};

module.exports = codeBlockPlugin;
