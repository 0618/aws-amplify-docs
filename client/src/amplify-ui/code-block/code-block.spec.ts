import {AmplifyCodeBlock} from "./code-block";
import {newSpecPage} from "@stencil/core/testing";

describe("amplify-code-block", () => {
  describe("Component logic", () => {
    let codeBlock: AmplifyCodeBlock;
    beforeEach(() => (codeBlock = new AmplifyCodeBlock()));

    it("should init `parsedLineCount` as `undefined`", () =>
      expect(codeBlock.parsedLineCount).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [AmplifyCodeBlock],
            html: `<amplify-code-block />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
