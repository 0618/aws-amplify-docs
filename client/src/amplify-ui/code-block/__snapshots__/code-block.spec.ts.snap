// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-code-block Render logic should render 1`] = `
<amplify-code-block class="css-1rchjef css-75wbz6 three-dee-effect">
  <!---->
  <div class="css-1e7418a"></div>
  <button class="css-tbb6d0">
    <span>
      copy
    </span>
  </button>
</amplify-code-block>
`;
